### DISCUSSION TITLE
What is this code doing?

### DISCUSSION THREAD

**Discussion by stevendarby** [dotnet/roslyn#68191] [Q&A] [answered]
```C#
var test = new Test1 { Test2 = { Value = "hello" } };

internal class Test1
{
    public Test2 Test2 { get; set; }
}

internal class Test2
{
    public string Value { get; set; }
}
```

I accidentally wrote some code like this, where I missed out `new Test2()` / `new()` when initializing the `Test2` property. I was surprised to find this compiles, but I don't really know why, and in what sense this code might be considered valid? It throws an exception at runtime:
```
System.NullReferenceException: 'Object reference not set to an instance of an object.'

Test1.Test2.get returned null.
```

**Comment by CyrusNajmabadi** [comment, marked as answer]
Yes.  this is entirely valid code.  `Test2 = { Value = "hello" }` is equivalent to writeing `Test2.Value = "Hello"` (which crashes in this case as expected).  There are cases where this would be totally fine.  For example, `Test2` might always initialize `Value` (or some subclass might, etc.).

**Reply by stevendarby** [reply]
Wow, I didn't know about this syntax. Thank you!

**Reply by CyrusNajmabadi** [reply]
You're welcome :)
